### DISCUSSION TITLE
How to pass an object from middleware to endpoint method?

### DISCUSSION THREAD

**Discussion by Gelembjuk** [fastapi/fastapi#12250] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import time

from fastapi import FastAPI, Request

app = FastAPI()

@app.middleware("http")
async def add_process_time_header(request: Request, call_next):
    myObject = MyClass(request)
    response = await call_next(request)
    
    return response

@app.get("/")
def read_root(arg:str):
    data = myObject.build_data(arg)
    return {"Hello": "World", "data": data}
```

### Description

I try to build a middleware for my app.

I need a way to create some object in my middleware function and i want to have access to that object in a next endpoint method . Please see my code example .

Is it possible to do? How to do this correctly?

Also, if it is not possible to create new object and pass it to an endpoint code. Maybe i can create kinda "global variable" for object? But how will it work in terms of multithreading? 

```
app = FastAPI()
myObject = MyClass(request)

@app.middleware("http")
async def add_process_time_header(request: Request, call_next):
    myObject.use_request(request)
    response = await call_next(request)
    
    return response

@app.get("/")
def read_root(arg:str):
    data = myObject.build_data(arg)
    return {"Hello": "World", "data": data}
```

Will this work fine or there will be some specifics of multithreading etc?

### Operating System

Linux

### FastAPI Version

0.115.0

### Pydantic Version

2.9.2

### Python Version

3.9

**Comment by sinisaos** [comment, marked as answer]
@Gelembjuk You can use [request.state](https://www.starlette.io/requests/#other-state) to pass data between middleware and FastAPI endpoints. Something like this
```python
from fastapi import FastAPI, Request


class MyClass:
    def __init__(self, arg: str) -> None:
        self.arg = arg


app = FastAPI()


@app.middleware("http")
async def add_process_time_header(request: Request, call_next):
    myObject = MyClass(arg="example argument")
    # add object to request state
    request.state.myObject = myObject
    response = await call_next(request)

    return response


@app.get("/")
def read_root(request: Request) -> dict:
    # use object from request state
    data = request.state.myObject
    return {"Hello": "World", "data": data.arg}
```

I hope this helps and sorry if I missed your point.

**Reply by Gelembjuk** [reply]
Thanks. This is exactly what i need
